add script to detect dead code

estute opened the pull request:
This adds a script to run vulture on the main code paths in the platform, excluding tests. At a later date, we may want to run this specifically *on* tests to clean up that section of the code.

One source of false positives that can arrise in the analysis is in templated html files. As far as I can see, there are three options for dealing with this:
* Accept that templates are a source of false positives and keep that in mind when doing dead code analysis
* Do some fancy grepping/python after running vulture to negate results when they are referenced in templates
* Look into ways of getting vulture to look at Mako/Jinja templates

I admit the third option would be a fun challenge, but odds are we want to do either the first or second option, due to the fact that we have no idea how often we plan on running this. Thoughts?

estute commented:
Another cause for false positives can be the fact that classes can be accessed via a string in settings/envs. Looking into how to deal with these

jmbowman commented on requirements/edx-sandbox/base.txt:
```diff
@@ -4,6 +4,7 @@
 #
 #    make upgrade
 #
+
```
I just filed [TE-2593](https://openedx.atlassian.net/browse/TE-2593) for this nuisance.  Could you make sure we don't switch the whitespace needlessly in this PR before that gets fixed?

estute replied:
Sure thing

estute commented on .gitignore:
```diff
@@ -138,3 +138,6 @@ dist
 .vscode
+
+# Vulture output
+dead-code.txt
```
Good point. I suppose I was thinking we would do something similar to the dependency script you wrote, but since the data generated from this script isn't fed into another tool, it does feel more like a report. I'll change it

estute commented:
I have also introduced another helper script (scripts/vulture/filter-vulture-results.sh). This iterates through the vulture script report and checks if there are any tokens that vulture has reported as dead that have only 1 instance within the codebase. This is helpful for looking for the low-hanging fruit, i.e. code that has no tests, is not used in templates, etc. Again, this is not fool proof, and many false positives still exist via dynamic python/django functionality, use of APIs and django signals.

Still, I was able to locate two pieces of dead code (I also searched the entire edx org for references and found none). Jiras created here:
* https://openedx.atlassian.net/browse/TE-2599
* https://openedx.atlassian.net/browse/TE-2600

jmbowman commented:
I'm not sure how useful that filter script will be in practice, although I suppose it's worth trying and/or refining further.  For example, the [test coverage](https://build.testeng.edx.org/view/edx-platform-master-tests/job/edx-platform-python-unittests-master/coveragepy/) indicates that the functions from TE-2600 are actually called; Django forms use methods named `clean_<fieldname>` in [validation](https://docs.djangoproject.com/en/1.11/ref/forms/validation/).  The one in TE-2599 seems legitimately unused, though.

Maybe it makes sense to bump `MIN_CONFIDENCE` up to 100 at first, if the false positives are drowning out the actual issues?

estute commented:
@jmbowman (un)fortunately, there are no instances of code that vulture is 100% confident about. I have removed the filter script, and moved `MIN_CONFIDENCE` to 90. I'll close TE-2600, but do you think we would be good to merge this one?

jmbowman commented:
Sure, this is a good start and easily covers the original ticket scope.
